### PR TITLE
TRUS-2664 Tax liability journey

### PR DESCRIPTION
**Pull request by adamconder**

### CODE REVIEW

File: app/models/TaxYearRange.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  private val start = uk.gov.hmrc.time.TaxYear.current.back(taxYear.year).starts.toString(fullDatePattern)
+  private val end = uk.gov.hmrc.time.TaxYear.current.back(taxYear.year).finishes.toString(fullDatePattern)
+
+  private val startingYearForTaxYear: String = uk.gov.hmrc.time.TaxYear.current.back(taxYear.year).startYear.toString
+
+  def yearAtStart: String = startingYearForTaxYear
+
+  def startYear: String = start
+
+  def endYear: String = end
```
Comment (jackheslop96) [inline app/models/TaxYearRange.scala:34]: Is it necessary to double up on defs/vals here? Feel free to ignore if that's good practice

File: app/services/TaxLiabilityService.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        TaxLiabilityYear(oldestYearToShow, earlierYears = true)
+      } else {
+        TaxLiabilityYear(taxYearOfDeath, earlierYears = false)
+      }
+    }
+  }
+
+  def getIsTaxLiabilityLate(taxYear: TaxYear, alreadyPaid: Boolean): Boolean = {
+    if ((!(localDateService.now.isBefore(currentTaxYearStartDate) || localDateService.now.isAfter(decemberDeadline))) &&
+      (taxYear.startYear == TaxYear.current.previous.startYear)) {
```
Comment (jackheslop96) [inline app/services/TaxLiabilityService.scala:78]: Can these be assigned to some boolean vals to make it a bit more readable?

File: app/models/TaxYearRange.scala (hunk 1 above)
Comment (adamconder) [reply app/models/TaxYearRange.scala:34]: I just wanted to hide the implementation, it's very java but it means we can also do additional logic without changing the value.

File: app/services/TaxLiabilityService.scala (hunk 2 above)
Comment (adamconder) [reply app/services/TaxLiabilityService.scala:78]: Agreed, missed this one! :D

File: test/connectors/EstatesStoreConnectorSpec.scala [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+class EstatesStoreConnectorSpec extends SpecBase
+  with ScalaFutures
+  with IntegrationPatience
+  with WireMockHelper {
+
+  implicit lazy val hc: HeaderCarrier = HeaderCarrier()
+
+  "trusts store connector" must {
+
```
Comment (jassalrichy) [inline test/connectors/EstatesStoreConnectorSpec.scala:36]: label as "Estates store connector"

File: app/controllers/CYMinusFourEarlierYearsLiabilityController.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import config.annotations.TaxLiability
+import controllers.actions.Actions
+import forms.YesNoFormProvider
```
Comment (jassalrichy) [inline app/controllers/CYMinusFourEarlierYearsLiabilityController.scala:21]: YesNoFormProvider import not required

File: app/controllers/CYMinusFourLiabilityController.scala [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import config.annotations.TaxLiability
+import controllers.actions.Actions
+import forms.{YesNoFormProvider, YesNoFormProviderWithArguments}
```
Comment (jassalrichy) [inline app/controllers/CYMinusFourLiabilityController.scala:21]: YesNoFormProvider import not required

File: app/controllers/CYMinusThreeEarlierYearsLiabilityController.scala [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import config.annotations.TaxLiability
+import controllers.actions.Actions
+import forms.YesNoFormProvider
```
Comment (jassalrichy) [inline app/controllers/CYMinusThreeEarlierYearsLiabilityController.scala:21]: YesNoFormProvider import not required

File: app/controllers/CYMinusThreeLiabilityController.scala [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import config.annotations.TaxLiability
+import controllers.actions.Actions
+import forms.{YesNoFormProvider, YesNoFormProviderWithArguments}
```
Comment (jassalrichy) [inline app/controllers/CYMinusThreeLiabilityController.scala:21]: YesNoFormProvider import not required

File: app/controllers/CYMinusTwoLiabilityController.scala [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import config.annotations.TaxLiability
+import controllers.actions.Actions
+import forms.{YesNoFormProvider, YesNoFormProviderWithArguments}
```
Comment (jassalrichy) [inline app/controllers/CYMinusTwoLiabilityController.scala:21]: YesNoFormProvider import not required

File: app/forms/YesNoFormProvider.scala [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+import javax.inject.Inject
+import play.api.data.Form
+
+class YesNoFormProvider @Inject() extends Mappings {
+
+  def withPrefix(prefix: String): Form[Boolean] =
+    Form(
+      "value" -> boolean(s"$prefix.error.required")
+    )
+}
```
Comment (jassalrichy) [inline app/forms/YesNoFormProvider.scala:29]: YesNoFormProvider not used ... can be deleted ?

File: app/utils/CheckYourAnswersHelper.scala [hunk 10]
```diff
@@ -16,27 +16,118 @@
...
+  private def yesNoPageForTaxYear(taxYear: TaxYear) : QuestionPage[Boolean] = taxYear match {
+    case CYMinus4TaxYear => CYMinusFourYesNoPage
+    case CYMinus3TaxYear => CYMinusThreeYesNoPage
+    case CYMinus2TaxYear => CYMinusTwoYesNoPage
+    case CYMinus1TaxYear => CYMinusOneYesNoPage
+  }
+
 }
 
 object CheckYourAnswersHelper {
```
Comment (jassalrichy) [inline app/utils/CheckYourAnswersHelper.scala:131]: CheckYourAnswersHelper object def empty .. can be removed ?

File: app/views/CYMinusFourYesNoView.scala.html [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    title = s"${errorPrefix(form)} ${messages("cyMinusFour.liability.title", taxYear)}"
+) {
+
+    @formHelper(action = CYMinusFourLiabilityController.onSubmit(mode), 'autoComplete -> "off") {
+
+        @components.back_link()
+
+        @components.error_summary(form.errors)
+
+        <h1 class="heading-large">@messages("cyMinusFour.liability.heading", taxYear)</h1>
```
Comment (jassalrichy) [inline app/views/CYMinusFourYesNoView.scala.html:38]: Does not follow the design pattern for radio options. i.e. H1 should be within legend: 
https://design-system.service.gov.uk/components/radios/
Applies for all YesNo views in PR.

Comment (BriWak) [reply app/views/CYMinusFourYesNoView.scala.html:38]: When there's content between the H1 and radio buttons it should just have the H1 text repeated in the fieldset and hidden. Adam has confirmed this with DAC

**Pull request merged**